File: server/src/main/kotlin/cn/enaium/zensquare/controller/ConversationController.kt
/*
 * ZenSquare is an opensource forums
 *
 * Copyright (C) 2023  Enaium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package cn.enaium.zensquare.controller

import cn.enaium.zensquare.bll.service.ConversationService
import cn.enaium.zensquare.model.entity.Thread
import cn.enaium.zensquare.model.entity.fetcher.ThreadFetcher
import cn.enaium.zensquare.model.entity.input.ThreadInput
import org.babyfish.jimmer.client.FetchBy
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import java.util.*

/**
 * @author Enaium
 */
@RestController
class ConversationController(
    val conversationService: ConversationService
) {

    /**
     * Create a conversation
     *
     * @param threadInput thread input
     * @param members members
     */
    @PutMapping("/threads/conversations/")
    @ResponseStatus(HttpStatus.OK)
    fun saveConversations(@RequestBody threadInput: ThreadInput, @RequestParam members: Array<UUID>) {
        conversationService.createConversation(threadInput, members)
    }

    /**
     * Find a conversation
     *
     * @param threadId thread id
     * @return conversation
     */
    @GetMapping("/members/threads/conversations/{threadId}")
    fun findConversation(@PathVariable threadId: UUID):
            @FetchBy("FULL_CONVERSATION", ownerType = ThreadFetcher::class) Thread? {
        return conversationService.findConversation(threadId)
    }

    /**
     * Find conversations
     *
     * @param page page
     * @param size size
     * @param memberId member id
     * @return conversations
     */
    @GetMapping("/members/{memberId}/threads/conversations/")
    fun findConversations(
        @RequestParam(defaultValue = "0") page: Int = 0,
        @RequestParam(defaultValue = "10") size: Int = 10,
        @PathVariable memberId: UUID
    ): Page<@FetchBy("DEFAULT_THREAD", ownerType = ThreadFetcher::class) Thread> {
        return conversationService.findConversations(PageRequest.of(page, size), memberId)
    }
}

File: website/src/__generated/services/ConversationController.ts
import type { Executor } from '../';
import type { ThreadDto } from '../model/dto';
import type { Page, ThreadInput, Unit } from '../model/static';

export class ConversationController {
    
    constructor(private executor: Executor) {}
    
    async findConversation(options: ConversationControllerOptions['findConversation']): Promise<
        ThreadDto['ThreadFetcher/FULL_CONVERSATION'] | undefined
    > {
        let _uri = '/members/threads/conversations/';
        _uri += encodeURIComponent(options.threadId);
        return (await this.executor({uri: _uri, method: 'GET'})) as ThreadDto['ThreadFetcher/FULL_CONVERSATION'] | undefined
    }
    
    async findConversations(options: ConversationControllerOptions['findConversations']): Promise<
        Page<ThreadDto['ThreadFetcher/DEFAULT_THREAD']>
    > {
        let _uri = '/members/';
        _uri += encodeURIComponent(options.memberId);
        _uri += '/threads/conversations/';
        let _separator = _uri.indexOf('?') === -1 ? '?' : '&';
        let _value: any = undefined;
        _value = options.page;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'page='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.size;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'size='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        return (await this.executor({uri: _uri, method: 'GET'})) as Page<ThreadDto['ThreadFetcher/DEFAULT_THREAD']>
    }
    
    async saveConversations(options: ConversationControllerOptions['saveConversations']): Promise<
        Unit
    > {
        let _uri = '/threads/conversations/';
        let _separator = _uri.indexOf('?') === -1 ? '?' : '&';
        let _value: any = undefined;
        _value = options.members.join(',');
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'members='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        return (await this.executor({uri: _uri, method: 'PUT', body: options.body})) as Unit
    }
}

export type ConversationControllerOptions = {
    'findConversation': {readonly threadId: string},
    'findConversations': {
        readonly page?: number, 
        readonly size?: number, 
        readonly memberId: string
    },
    'saveConversations': {readonly body: ThreadInput, readonly members: ReadonlyArray<string>}
}